src/backend/distributed/commands/dependencies.c

```diff
@@ -0,0 +1,196 @@
+	if (list_length(workerNodeList) <= 0)
+	{
+		/* no nodes to execute on */
+		return;
+	}
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		MultiConnection *connection = NULL;
+
+		char *nodeName = workerNode->workerName;
+		uint32 nodePort = workerNode->workerPort;
+
+		connection = GetNodeUserDatabaseConnection(connectionFlag, nodeName,
+												   nodePort,
+												   CitusExtensionOwnerName(),
+												   NULL);
+
+		connections = lappend(connections, connection);
+	}
+
+	/*
+	 * create dependency on all nodes
+	 */
+	foreach(connectionCell, connections)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		ExecuteCriticalRemoteCommandList(connection, ddlCommands);
+	}
+
+	/*
+	 * mark all objects as distributed
+	 */
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		markObjectDistributed(dependency);
+	}
+
+	/*
+	 * disconnect from nodes
+	 */
+	foreach(connectionCell, connections)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		CloseConnection(connection);
+	}
+}
+
+
+/*
+ * GetDependencyCreateDDLCommands returns a list (potentially empty or NIL) of ddl
+ * commands to execute on a worker to create the object.
+ */
+static List *
+GetDependencyCreateDDLCommands(const ObjectAddress *dependency)
+{
+	switch (getObjectClass(dependency))
+	{
+		case OCLASS_SCHEMA:
+		{
+			const char *schemaDDLCommand = CreateSchemaDDLCommand(dependency->objectId);
+
+			if (schemaDDLCommand == NULL)
+			{
+				/* no schema to create */
+				return NIL;
+			}
+
+			return list_make1((void *) schemaDDLCommand);
+		}
+
+		default:
+		{
+			return NIL;
+		}
+	}
+}
+
+
+/*
+ * ReplicateAllDependenciesToNode replicate all previously marked objects to a worker node
+ */
+void
+ReplicateAllDependenciesToNode(const char *nodeName, int nodePort)
+{
+	const uint64 connectionFlag = FORCE_NEW_CONNECTION;
+	ListCell *dependencyCell = NULL;
+	List *dependencies = NIL;
+	List *ddlCommands = NIL;
+	MultiConnection *connection = NULL;
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	dependencies = GetDistributedObjectAddressList();
+	dependencies = OrderObjectAddressListInDependencyOrder(dependencies);
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		ddlCommands = list_concat(ddlCommands,
+								  GetDependencyCreateDDLCommands(dependency));
+	}
+	if (list_length(ddlCommands) <= 0)
+	{
+		/* no commands to replicate dependencies to the new worker */
+		return;
+	}
+
+	/*
+	 * connect to the new host and create all applicable dependencies
+	 */
+	connection = GetNodeUserDatabaseConnection(connectionFlag, nodeName, nodePort,
+											   CitusExtensionOwnerName(), NULL);
+	ExecuteCriticalRemoteCommandList(connection, ddlCommands);
```

Comment:
Similar here, with many dependencies, these round-trips per type becomes the bottleneck.

Reply:
wouldn't it make sense to solve this in `ExecuteCriticalRemoteCommandList`? guess more places could benefit from this.

Reply:
But then some callers might be relying on that behaviour? As I noted in another comment, I think it is not very critical, we can discuss that later.